src/System.Net.Sockets/tests/FunctionalTests/SendReceive.cs

```diff
@@ -1334,6 +1336,114 @@ select Task.Factory.StartNew(() => pair.Item1.Receive(new byte[1]), Cancellation
         }
     }
 
+    public class BlockingSendReceive : RemoteExecutorTestBase
+    {
+
+        private readonly ITestOutputHelper _log;
+        public static readonly object[][] Loopbacks = MemberDatas.Loopbacks;
+
+        public BlockingSendReceive(ITestOutputHelper output)
+        {
+            _log = output;
+        }
+
+        [Theory]
+        [MemberData(nameof(Loopbacks))]
+        public void BlockingStreamRead_Close(IPAddress ipAddress)
+        {
+            // This test verifies blocking behavior. Always run it as remote task
+            // to isolate any possible failures.
+            RemoteInvoke((address) =>
+            {
+                IPAddress listenAt = IPAddress.Parse(address);
+
+                // Test for #22564 when close is called while blocking read is pending.
+                using (var server = new Socket(listenAt.AddressFamily, SocketType.Stream, ProtocolType.Tcp))
+                using (var client = new Socket(listenAt.AddressFamily, SocketType.Stream, ProtocolType.Tcp))
+                {
+                    server.BindToAnonymousPort(listenAt);
+                    server.Listen(1);
+                    byte[] buffer = new byte[1];
+                    buffer[0] = Convert.ToByte('a');
+
+                    Thread clientThread = new Thread(() =>
+                    {
+                        client.Connect(server.LocalEndPoint);
+                        client.Send(buffer);
+                        // Blocking read.
+                        try
+                        {
+                            client.Receive(buffer);
+                        }
+                        catch (SocketException) { }
+                        catch (ObjectDisposedException) { }
+                    });
+                    clientThread.IsBackground = true;
+                    clientThread.Start();
+                    Socket s = server.Accept();
+                    // Try to receive something so we know client thread is running.
+                    s.Receive(buffer, 1, SocketFlags.None);
+                    // There is however race condition that if the Receive() is not started yet, this will simply
+                    // finish without hitting the test scenario.
+                    Thread.Sleep(500);
+
+                    // clientThread should be now blocked on Receive()
+                    // Close client socket from parent thread.
+                    client.Close();
+                    clientThread.Join();
+                }
+                return SuccessExitCode;
+            }, ipAddress.ToString()).Dispose();
+        }
+
+        [Theory]
+        [PlatformSpecific(~TestPlatforms.OSX)]
+        [MemberData(nameof(Loopbacks))]
+        public void BlockingStreamAccept_Close(IPAddress ipAddress)
+        {
+            // This test verifies blocking behavior. Always run it as remote task
+            // to isolate any possible failures.
+            RemoteInvoke((address) =>
```

Comment:
why isn't thread enough isolation?

Reply:
Guess: failure could manifest as exception on threadpool thread?

Reply:
Isn't Task.Run going to take care of all of the exceptions softly? Could you give example (or simulation) of failure which we are afraid of?

Reply:
I wanted to be sure that if something goes wrong we do not block whole xunit suite @krwq.
RemoteProcess seems more resilient. (even if maybe harder to debug)

Reply:
I agree with @krwq here. If the test is failing, it doesn't really matter whether it is blocking the rest of xunit or not. Adding RemoteInvoke just seems to make the test more complicated.